pygmt/xarray_backend.py

```diff
@@ -0,0 +1,106 @@
+from pygmt.clib import Session
+from pygmt.exceptions import GMTInvalidInput
+from pygmt.helpers import build_arg_list
+from pygmt.src.which import which
+from xarray.backends import BackendEntrypoint
+
+
+class GMTBackendEntrypoint(BackendEntrypoint):
+    """
+    Xarray backend to read raster grid/image files using 'gmt' engine.
+
+    Internally, GMT uses the NetCDF C library to read NetCDF files, and GDAL for GeoTIFF
+    and other raster formats.
+
+    When using :py:func:`xarray.open_dataarray` or :py:func:`xarray.load_dataarray` with
+    ``engine="gmt"``, pass the ``raster_kind`` parameter that can be either:
+
+    - ``"grid"`` - for reading single-band raster grids
+    - ``"image"`` - for reading multi-band raster images
+
+    Examples
+    --------
+    Read a single-band NetCDF file using ``raster_kind="grid"``
+
+    >>> import pygmt
+    >>> import xarray as xr
+    >>>
+    >>> da_grid = xr.open_dataarray(
+    ...     "@static_earth_relief.nc", engine="gmt", raster_kind="grid"
+    ... )
+    >>> da_grid  # doctest: +NORMALIZE_WHITESPACE, +ELLIPSIS
+    <xarray.DataArray 'z' (lat: 14, lon: 8)>...
+    [112 values with dtype=float32]
+    Coordinates:
+      * lat      (lat) float64... -23.5 -22.5 -21.5 -20.5 ... -12.5 -11.5 -10.5
+      * lon      (lon) float64... -54.5 -53.5 -52.5 -51.5 -50.5 -49.5 -48.5 -47.5
+    Attributes:
+        Conventions:   CF-1.7
+        title:         Produced by grdcut
+        history:       grdcut @earth_relief_01d_p -R-55/-47/-24/-10 -Gstatic_eart...
+        description:   Reduced by Gaussian Cartesian filtering (111.2 km fullwidt...
+        actual_range:  [190. 981.]
+        long_name:     elevation (m)
+
+    Read a multi-band GeoTIFF file using ``raster_kind="image"``
+
+    >>> da_image = xr.open_dataarray(
+    ...     "@earth_night_01d", engine="gmt", raster_kind="image"
+    ... )
+    >>> da_image  # doctest: +NORMALIZE_WHITESPACE, +ELLIPSIS
+    <xarray.DataArray 'z' (band: 3, y: 180, x: 360)>...
+    [194400 values with dtype=uint8]
+    Coordinates:
+      * y        (y) float64... 89.5 88.5 87.5 86.5 ... -86.5 -87.5 -88.5 -89.5
+      * x        (x) float64... -179.5 -178.5 -177.5 -176.5 ... 177.5 178.5 179.5
+      * band     (band) uint8... 1 2 3
+    Attributes:
+        long_name:  z
+    """
+
+    description = "Open raster (.grd, .nc or .tif) files in Xarray via GMT."
+    open_dataset_parameters = ("filename_or_obj", "raster_kind")
+    url = "https://pygmt.org/dev/api/generated/pygmt.GMTBackendEntrypoint.html"
+
+    def open_dataset(  # type: ignore[override]
+        self,
+        filename_or_obj: PathLike,
+        *,
+        drop_variables=None,  # noqa: ARG002
+        raster_kind: Literal["grid", "image"],
+        # other backend specific keyword arguments
+        # `chunks` and `cache` DO NOT go here, they are handled by xarray
+    ) -> xr.Dataset:
+        """
+        Backend open_dataset method used by Xarray in :py:func:`~xarray.open_dataset`.
```

Comment:
With the fix in #3927, sphinx-autogen will generate stub file for this method. It's not ideal, and we need add a "Parameters" section at least.